### PR TITLE
Upgrade to PHP 7

### PR DESCRIPTION
**Pull request by TomasVotruba**
Replaces #647 with passing tests and CS

## Why?

- majority of PHP [world bumping to PHP 7.1](https://gophp71.org/)
- upcoming [PHP 7.2](https://github.com/php-ai/php-ml/pull/147)
- we could also add PHP 7.1 typehints via coding standards], instead of doing that twice to just PHP 7.0
- [37 % people](https://seld.be/notes/php-versions-stats-2017-2-edition) are on PHP 7.1 and growing

This  package would be [still available for PHP 5.6-](https://seld.be/images/update-reqs.png).

### 1. How to approach `string`/`int` issue in parameters?

Often there was random typehint and for `string` and `int` there was autore-type to string `rawurlencode` method.

Should be allow both `string` and `int` and manuall `rawurlencode((string) $id)` or be strict in parameters?

### 2. How to approach `null`/`array` return values?

At least accroding to tests, some api calls return null on fail or empty results. Array was expected in typehtings.

At the moment they return `array|null`. Is that ok?

### CODE REVIEW

**Comment by m1guelpf** [comment]
> How to approach null/array return values?

@TomasVotruba I'm not really sure, but I think in PHP7 you can use `array?` as a typehint, which I believe is better

**Comment by TomasVotruba** [comment]
@m1guelpf Sure. We should move to PHP 7.1 in standalone PR, oen step at a time.

This is rather architecture question, whether all these methods should return array/null

**Comment by GrahamCampbell** [comment]
> Sure. We should move to PHP 7.1 in standalone PR, oen step at a time.

I thin that's probably a bit fast. `^7.0` will be a good version constraint through 2018 I think.

**Comment by TomasVotruba** [comment]
@GrahamCampbell My point was *not in this PR. It should be PHP 7.0 only*

**Comment by GrahamCampbell** [comment]
Sure. I was just making the point that I'd 👎 anything beyond this PR unless there were some really good reasons.

**Comment by cursedcoder** [comment]
What about strict types declaration in every file?
Is it possible to enforce it once? Maybe some composer flag?

**Comment by TomasVotruba** [comment]
@cursedcoder Sure, we can require that in coding standard, as I did in [php-ml](https://github.com/php-ai/php-ml/pull/156/files#diff-db71ac2c5cb10a341e624322fac72c1e). But again, I'd do that in separate PR, after this one gets resolved to prevent redundant conflicts.

**Comment by Nyholm** [comment]
I will tag 2.7 within a few minutes. 
Could you resolve these conflicts and then merge this PR and make 2.8 our first version without PHP5.

**Comment by TomasVotruba** [comment]
Hey, I'm glad you got to this.

Sure, let me know after 2.7 is tagged

**Comment by Nyholm** [comment]
It's tagged now =)

**Comment by TomasVotruba** [comment]
Rebase done :+1:

**Comment by TomasVotruba** [comment]
Btw, what about bumping to PHP 7.1? Even [PHPUnit is now on board](https://gophp71.org)

File: lib/Github/Api/Authorizations.php [hunk 1]
```diff
@@ -12,110 +12,87 @@ class Authorizations extends AbstractApi
...
 
     /**
      * Show a single authorization.
-     *
-     * @param $clientId
-     *
-     * @return array
      */
-    public function show($clientId)
+    public function show($clientId): array
```
Comment (Nyholm) [inline lib/Github/Api/Authorizations.php]: should be a string here, right?

File: lib/Github/Api/Authorizations.php [hunk 2]
```diff
@@ -12,110 +12,87 @@ class Authorizations extends AbstractApi
...
     {
-        return $this->get('/authorizations/'.rawurlencode($clientId));
+        return $this->get('/authorizations/'.rawurlencode((string) $clientId));
```
Comment (Nyholm) [inline lib/Github/Api/Authorizations.php]: Type casting here could be removed.

File: lib/Github/Api/Authorizations.php [hunk 3]
```diff
@@ -12,110 +12,87 @@ class Authorizations extends AbstractApi
...
     /**
      * Create an authorization.
      *
-     * @param array $params
      * @param null $OTPCode
      *
-     * @return array
+     * @return array|null
      */
     public function create(array $params, $OTPCode = null)
```
Comment (Nyholm) [inline lib/Github/Api/Authorizations.php]: $OPTCode is string, right?

File: lib/Github/Api/Authorizations.php [hunk 4]
```diff
@@ -12,110 +12,87 @@ class Authorizations extends AbstractApi
...
-     * @param $clientId
-     * @param array $params
-     *
-     * @return array
+     * @return array|null
      */
     public function update($clientId, array $params)
     {
-        return $this->patch('/authorizations/'.rawurlencode($clientId), $params);
+        return $this->patch('/authorizations/'.rawurlencode((string) $clientId), $params);
```
Comment (Nyholm) [inline lib/Github/Api/Authorizations.php]: Same here, shouldnt client ID be a string? Can you verify?

File: lib/Github/Api/Authorizations.php [hunk 5]
```diff
@@ -12,110 +12,87 @@ class Authorizations extends AbstractApi
...
     {
-        return $this->get('/applications/'.rawurlencode($clientId).'/tokens/'.rawurlencode($token));
+        return $this->get('/applications/'.rawurlencode((string) $clientId).'/tokens/'.rawurlencode($token));
     }
 
     /**
      * Reset an authorization.
      *
-     * @param $clientId
-     * @param $token
```
Comment (Nyholm) [inline lib/Github/Api/Authorizations.php]: We should not remove these with adding type annotations.

**Comment by TomasVotruba** [comment]
Every added type was to fix tests, that failed due to strict types.

File: lib/Github/Api/Authorizations.php (hunk 5 above)
Comment (TomasVotruba) [reply lib/Github/Api/Authorizations.php]: What is the type here?

File: lib/Github/Exception/ErrorException.php [hunk 6]
```diff
@@ -7,6 +7,6 @@
  *
  * @author Joseph Bielawski <stloyd@gmail.com>
  */
-class ErrorException extends \ErrorException implements ExceptionInterface
+class ErrorException extends \Exception implements ExceptionInterface
```
Comment (acrobat) [inline lib/Github/Exception/ErrorException.php]: Why is this change needed?

Comment (TomasVotruba) [reply lib/Github/Exception/ErrorException.php]: Not sure anymore, it's quite a long time.

I think it was related to mocks in some way. Try running test with/without it please to confirm.

**Comment by acrobat** [comment]
Imho this PR does a bit to much. Bumping php, adding typehints, cs fixes, etc. This makes it a bit hard for **me** to review the PR. Maybe we should split this up a bit. 
Leave the style changes out for example as this can be done in a separate pr (I've started one with new styleci fixers, but maybe more is needed). 
Bump the php version, adapt travis to it and make sure the tests still pass. After that we can start converting typehints. This way we have multiple "small" pr's which makes easy reviews and quick merges :)

This is just my opinion of course, don't know what you guys think of this.

File: lib/Github/Api/Authorizations.php (hunk 5 above)
Comment (Nyholm) [reply lib/Github/Api/Authorizations.php]: I’m not sure without reading the code and docs.

**Comment by Nyholm** [comment]
I agree with acrobat. 

Tomas, what is the status here?

**Comment by TomasVotruba** [comment]
What is the decision here?

**Comment by acrobat** [comment]
I think we should split the pr. Leave this one for now and open a new PR where you bump the php version, adapt the travis build matrix and make sure tests still pass. From there on we can make the next steps (like adding typehints etc) in separate pr's.

File: lib/Github/Api/AbstractApi.php [hunk 7]
```diff
@@ -1,4 +1,4 @@
-<?php
+<?php declare(strict_types=1);
```
Comment (Nek-) [inline lib/Github/Api/AbstractApi.php]: I don't think this is in the way of think of the project. Keeping the PHP way looks like a better alternative. Also, the indentation is wrong (it should definitely be on a new line)

File: lib/Github/Api/ApiInterface.php [hunk 8]
```diff
@@ -11,5 +11,5 @@
 {
     public function getPerPage();
 
-    public function setPerPage($perPage);
+    public function setPerPage(int $perPage = null);
```
[the remark this answers is not in the document]
Comment (Nek-) [reply lib/Github/Api/ApiInterface.php]: This is not something new. You can specify no pagination and let github decide AFAIK. It's written as comment on some inheritance of this interface.

**Comment by Nek-** [comment]
About the first point (`1. How to approach string/int issue in parameters?`) the proper way to fix that is to not declare strict type hinting and type hinting `string` (the result is a cast).

**Comment by TomasVotruba** [comment]
Not relevant anymore. I'm closing to clean my opened PRs.

Feel free to cherry pick anything you need.
